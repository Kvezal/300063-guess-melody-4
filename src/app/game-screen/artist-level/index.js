import ArtistLevel from "./artist-level";


export default ArtistLevel;
