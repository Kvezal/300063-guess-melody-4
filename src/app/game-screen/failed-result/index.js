import FailedResult from "./failed-result";


export default FailedResult;
