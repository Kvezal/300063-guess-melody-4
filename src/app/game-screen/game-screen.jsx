import React, {PureComponent} from "react";
import PropTypes from "prop-types";
import {Link} from "react-router-dom";
import {connect} from "react-redux";

import Types from "@types";
import {GameScreenTypes, GameLevels} from "@enums";
import MistakeList from "@components/mistake-list";
import GenreLevel from "@app/game-screen/genre-level";
import ArtistLevel from "@app/game-screen/artist-level";
import FailedResult from "@app/game-screen/failed-result";
import SuccessResult from "@app/game-screen/success-result";
import withActivePlayer from "@hocs/with-active-player";

import {ActionCreator} from "../../reducer";

const GenreLevelWrapper = withActivePlayer(GenreLevel);
const ArtistLevelWrapper = withActivePlayer(ArtistLevel);


class GameScreen extends PureComponent {
  constructor(props) {
    super(props);

    this._chooseGameScreen = this._chooseGameScreen.bind(this);
    this._chooseSuccessResultScreen = this._chooseSuccessResultScreen.bind(this);
    this._chooseFailedResultScreen = this._chooseFailedResultScreen.bind(this);
    this._screenMap = new Map([
      [GameScreenTypes.SUCCESS, () => this._chooseSuccessResultScreen()],
      [GameScreenTypes.FAILED, () => this._chooseFailedResultScreen()],
      [GameScreenTypes.GAME, this._chooseGameScreen],
    ]);

    this._chooseArtistLevel = this._chooseArtistLevel.bind(this);
    this._chooseGenreLevel = this._chooseGenreLevel.bind(this);
    this._levelMap = new Map([
      [GameLevels.ARTIST, this._chooseArtistLevel],
      [GameLevels.GENRE, this._chooseGenreLevel]
    ]);
  }

  render() {
    const {screen} = this.props;
    return this._screenMap.get(screen)();
  }

  _chooseSuccessResultScreen() {
    const {level, mistakes, onResetGame} = this.props;
    return <SuccessResult
      rightAnswers={level - mistakes}
      mistakes={mistakes}
      onResetLinkClick={onResetGame}
    />;
  }

  _chooseFailedResultScreen() {
    const {onResetGame} = this.props;
    return <FailedResult onResetLinkClick={onResetGame}/>;
  }

  _chooseGameScreen() {
    const {questions, level, mistakes} = this.props;
    const question = questions[level];
    return <section className={`game game--${question.type}`}>
      <header className="game__header">
        <Link className="game__back" to="/">
          <span className="visually-hidden">Сыграть ещё раз</span>
          <img className="game__logo" src="img/melody-logo-ginger.png" alt="Угадай мелодию" />
        </Link>
        <svg xmlns="http://www.w3.org/2000/svg" className="timer" viewBox="0 0 780 780">
          <circle className="timer__line" cx="390" cy="390" r="370" />
        </svg>
        <MistakeList mistakes={mistakes} />
      </header>
      {this._levelMap.get(question.type)(question)}
    </section>;
  }

  _chooseArtistLevel(question) {
    const {onUserAnswer} = this.props;
    return <ArtistLevelWrapper
      question={question}
      onAnswer={(answer) => onUserAnswer(question, answer)}
    />;
  }

  _chooseGenreLevel(question) {
    const {onUserAnswer} = this.props;
    return <GenreLevelWrapper
      question={question}
      onAnswer={(answers) => onUserAnswer(question, answers)}
    />;
  }
}

GameScreen.propTypes = {
  questions: PropTypes.arrayOf(Types.question).isRequired,
  level: PropTypes.number.isRequired,
  mistakes: PropTypes.number.isRequired,
  screen: PropTypes.string.isRequired,
  onUserAnswer: PropTypes.func.isRequired,
  onResetGame: PropTypes.func.isRequired,
};

const mapStateToProps = (state) => ({
  questions: state.questions,
  level: state.level,
  mistakes: state.mistakes,
  screen: state.screen,
});

const mapDispatchToProps = (dispatch) => ({
  onUserAnswer(question, answer) {
    dispatch(ActionCreator.incrementLevel());
    dispatch(ActionCreator.incrementMistake(question, answer));
    dispatch(ActionCreator.chooseScreen());
  },

  onResetGame() {
    dispatch(ActionCreator.resetGame());
    dispatch(ActionCreator.chooseScreen());
  }
});

export {GameScreen};
export default connect(mapStateToProps, mapDispatchToProps)(GameScreen);
