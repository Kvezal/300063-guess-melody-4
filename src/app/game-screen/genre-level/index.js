import GenreLevel from "./genre-level";


export default GenreLevel;
