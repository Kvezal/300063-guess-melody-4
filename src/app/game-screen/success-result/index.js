import SuccessResult from "./success-result";


export default SuccessResult;
