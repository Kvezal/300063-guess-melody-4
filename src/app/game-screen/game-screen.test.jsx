import React from "react";
import render from "react-test-renderer";
import {MemoryRouter} from "react-router-dom";

import {GameScreen} from "./game-screen";


describe(`GameScreen`, () => {
  test(`should render component`, () => {
    const questions = [{
      type: `artist`,
      song: {artist: `Jim Beam`, src: `src`},
      answers: [
        {picture: `picture/1`, artist: `John Snow`},
        {picture: `picture/2`, artist: `Jack Daniels`},
        {picture: `picture/3`, artist: `Jim Beam`}
      ],
    }];
    const tree = render
      .create(
          <MemoryRouter>
            <GameScreen
              questions={questions}
              level={0}
              mistakes={0}
              screen="game"
              onUserAnswer={() => {}}
              onResetGame={() => {}}
            />
          </MemoryRouter>, {
            createNodeMock: () => ({}),
          }
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  test(`should render component`, () => {
    const questions = [{
      type: `genre`,
      genre: `rock`,
      answers: [
        {src: `src/1`, genre: `rock`},
        {src: `src/2`, genre: `blues`},
        {src: `src/3`, genre: `jazz`},
        {src: `src/4`, genre: `rock`}
      ],
    }];
    const tree = render
      .create(
          <MemoryRouter>
            <GameScreen
              questions={questions}
              level={0}
              mistakes={0}
              screen="game"
              onUserAnswer={() => {}}
              onResetGame={() => {}}
            />
          </MemoryRouter>, {
            createNodeMock: () => ({}),
          }
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
