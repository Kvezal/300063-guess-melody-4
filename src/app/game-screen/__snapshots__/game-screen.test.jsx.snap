// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GameScreen should render component 1`] = `
<section
  className="game game--artist"
>
  <header
    className="game__header"
  >
    <a
      className="game__back"
      href="/"
      onClick={[Function]}
    >
      <span
        className="visually-hidden"
      >
        Сыграть ещё раз
      </span>
      <img
        alt="Угадай мелодию"
        className="game__logo"
        src="img/melody-logo-ginger.png"
      />
    </a>
    <svg
      className="timer"
      viewBox="0 0 780 780"
      xmlns="http://www.w3.org/2000/svg"
    >
      <circle
        className="timer__line"
        cx="390"
        cy="390"
        r="370"
      />
    </svg>
    <div
      className="game__mistakes"
    />
  </header>
  <section
    className="game__screen"
  >
    <h2
      className="game__title"
    >
      Кто исполняет эту песню?
    </h2>
    <div
      className="game__track"
    >
      <div
        className="track"
      >
        <button
          className="track__button track__button--pause"
          onClick={[Function]}
          type="button"
        />
        <div
          className="track__status"
        >
          <audio />
        </div>
      </div>
    </div>
    <form
      className="game__artist"
    >
      <div
        className="artist"
      >
        <input
          className="artist__input visually-hidden"
          id="John Snow"
          name="answer"
          onChange={[Function]}
          type="radio"
          value="John Snow"
        />
        <label
          className="artist__name"
          htmlFor="John Snow"
        >
          <img
            alt="John Snow"
            className="artist__picture"
            src="picture/1"
          />
          John Snow
        </label>
      </div>
      <div
        className="artist"
      >
        <input
          className="artist__input visually-hidden"
          id="Jack Daniels"
          name="answer"
          onChange={[Function]}
          type="radio"
          value="Jack Daniels"
        />
        <label
          className="artist__name"
          htmlFor="Jack Daniels"
        >
          <img
            alt="Jack Daniels"
            className="artist__picture"
            src="picture/2"
          />
          Jack Daniels
        </label>
      </div>
      <div
        className="artist"
      >
        <input
          className="artist__input visually-hidden"
          id="Jim Beam"
          name="answer"
          onChange={[Function]}
          type="radio"
          value="Jim Beam"
        />
        <label
          className="artist__name"
          htmlFor="Jim Beam"
        >
          <img
            alt="Jim Beam"
            className="artist__picture"
            src="picture/3"
          />
          Jim Beam
        </label>
      </div>
    </form>
  </section>
</section>
`;

exports[`GameScreen should render component 2`] = `
<section
  className="game game--genre"
>
  <header
    className="game__header"
  >
    <a
      className="game__back"
      href="/"
      onClick={[Function]}
    >
      <span
        className="visually-hidden"
      >
        Сыграть ещё раз
      </span>
      <img
        alt="Угадай мелодию"
        className="game__logo"
        src="img/melody-logo-ginger.png"
      />
    </a>
    <svg
      className="timer"
      viewBox="0 0 780 780"
      xmlns="http://www.w3.org/2000/svg"
    >
      <circle
        className="timer__line"
        cx="390"
        cy="390"
        r="370"
      />
    </svg>
    <div
      className="game__mistakes"
    />
  </header>
  <section
    className="game__screen"
  >
    <h2
      className="game__title"
    >
      Выберите инди-рок треки
    </h2>
    <form
      className="game__tracks"
      onSubmit={[Function]}
    >
      <div
        className="track"
      >
        <button
          className="track__button track__button--pause"
          onClick={[Function]}
          type="button"
        />
        <div
          className="track__status"
        >
          <audio />
        </div>
        <div
          className="game__answer"
        >
          <input
            className="game__input visually-hidden"
            id="rock0"
            name="answer"
            onChange={[Function]}
            type="checkbox"
            value="rock"
          />
          <label
            className="game__check"
            htmlFor="rock0"
          >
            Отметить
          </label>
        </div>
      </div>
      <div
        className="track"
      >
        <button
          className="track__button track__button--play"
          onClick={[Function]}
          type="button"
        />
        <div
          className="track__status"
        >
          <audio />
        </div>
        <div
          className="game__answer"
        >
          <input
            className="game__input visually-hidden"
            id="blues1"
            name="answer"
            onChange={[Function]}
            type="checkbox"
            value="blues"
          />
          <label
            className="game__check"
            htmlFor="blues1"
          >
            Отметить
          </label>
        </div>
      </div>
      <div
        className="track"
      >
        <button
          className="track__button track__button--play"
          onClick={[Function]}
          type="button"
        />
        <div
          className="track__status"
        >
          <audio />
        </div>
        <div
          className="game__answer"
        >
          <input
            className="game__input visually-hidden"
            id="jazz2"
            name="answer"
            onChange={[Function]}
            type="checkbox"
            value="jazz"
          />
          <label
            className="game__check"
            htmlFor="jazz2"
          >
            Отметить
          </label>
        </div>
      </div>
      <div
        className="track"
      >
        <button
          className="track__button track__button--play"
          onClick={[Function]}
          type="button"
        />
        <div
          className="track__status"
        >
          <audio />
        </div>
        <div
          className="game__answer"
        >
          <input
            className="game__input visually-hidden"
            id="rock3"
            name="answer"
            onChange={[Function]}
            type="checkbox"
            value="rock"
          />
          <label
            className="game__check"
            htmlFor="rock3"
          >
            Отметить
          </label>
        </div>
      </div>
      <button
        className="game__submit button"
        type="submit"
      >
        Ответить
      </button>
    </form>
  </section>
</section>
`;
